### DISCUSSION TITLE
How do you return json and specify exit code?

### DISCUSSION THREAD

**Discussion by jasonkuhrt** [oclif/core#1270] [Q&A] [answered]
We're adding JSON support to our CLI. This requires us to return data from commands. I'm struggling in a few places to reconcile how that squares with exiting the process with non-zero codes. Can anyone provide some guidance on this pattern?

```ts
        if (forceSafe) {
          this.logSuccess('Breaking changes were expected (forced)');
          return this.successData({
            data: {
              checkType: 'registry',
              breakingChanges: true,
              warnings: toWaring(result.schemaCheck.warnings),
              changes: toChange(result.schemaCheck.changes),
              url: result.schemaCheck?.schemaCheck?.webUrl ?? null,
            },
          });
        }

        // TODO integrate return JSON with exit 1
        this.exit(1);
      }
```

**Comment by jasonkuhrt** [comment]
I'm currently exploring creating a subclass of [ExitError](https://github.com/oclif/core/blob/main/src/errors/errors/exit.ts) that would output given data as json.

**Comment by jasonkuhrt** [comment]
An even simpler solution my be just `process.exitCode = 1;`.

**Comment by mdonnalley** [comment]
If you want to modify the JSON that's returned when the command errors (or you call `this.exit` or `this.error`), you should override the `toErrorJson` method: https://github.com/oclif/core/blob/4.0.37/src/command.ts#L296

**Reply by jasonkuhrt** [reply]
Thanks @mdonnalley, will look more closely at that.

**Comment by jasonkuhrt** [comment]
I would like a way to modify the exit code without using `throw`. What would you suggest to achieve that?

**Reply by mdonnalley** [reply, marked as answer]
you can use `process.exitCode = 9999` and oclif should respect that up when exiting the command
